pkg/compact/blocks_cleaner.go

```diff
@@ -0,0 +1,53 @@
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
+package compact
+
+import (
+	"context"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/thanos-io/thanos/pkg/block"
+	"github.com/thanos-io/thanos/pkg/objstore"
+)
+
+// BlocksCleaner is a struct that deletes blocks from bucket which are marked for deletion.
+type BlocksCleaner struct {
+	logger                   log.Logger
+	ignoreDeletionMarkFilter *block.IgnoreDeletionMarkFilter
+	bkt                      objstore.Bucket
+	deleteDelay              time.Duration
+	blocksCleaned            prometheus.Counter
+}
+
+// NewBlocksCleaner creates a new BlocksCleaner.
+func NewBlocksCleaner(logger log.Logger, bkt objstore.Bucket, ignoreDeletionMarkFilter *block.IgnoreDeletionMarkFilter, deleteDelay time.Duration, blocksCleaned prometheus.Counter) *BlocksCleaner {
+	return &BlocksCleaner{
+		logger:                   logger,
+		ignoreDeletionMarkFilter: ignoreDeletionMarkFilter,
+		bkt:                      bkt,
+		deleteDelay:              deleteDelay,
+		blocksCleaned:            blocksCleaned,
+	}
+}
+
+// DeleteMarkedBlocks uses ignoreDeletionMarkFilter to delete the blocks that are marked for deletion.
+func (s *BlocksCleaner) DeleteMarkedBlocks(ctx context.Context) error {
+	level.Debug(s.logger).Log("msg", "started cleaning of blocks marked for deletion")
+
+	toDeleteBlockIds := s.ignoreDeletionMarkFilter.DeletionMarkBlocks(s.deleteDelay)
+	for _, id := range toDeleteBlockIds {
+		if err := block.Delete(ctx, s.logger, s.bkt, id); err != nil {
+			return errors.Wrap(err, "delete block")
+		}
+		s.blocksCleaned.Inc()
```

Comment:
It would be nice to have metric for failures.. but maybe bucket metric for op `delete` is good enough

Reply:
Added since its nice to have. Resolving this thread.